src/main/core/scheduler/scheduler.c

```diff
@@ -228,20 +228,37 @@ void scheduler_unref(Scheduler* scheduler) {
     }
 }
 
+static void _scheduler_unref_event(Host* host, void* voidEvent, void* unused) {
+    Event* event = voidEvent;
+    event_unref(event);
+}
+
 gboolean scheduler_push(Scheduler* scheduler, Event* event, Host* sender, Host* receiver) {
     MAGIC_ASSERT(scheduler);
 
-    SimulationTime eventTime = event_getTime(event);
-    if(eventTime >= scheduler->endTime) {
-        event_unref(event);
-        return FALSE;
-    }
-
     /* parties involved. sender may be NULL, receiver may not!
      * we MAY NOT OWN the receiver, so do not write to it! */
     utility_assert(receiver);
     utility_assert(receiver == event_getHost(event));
 
+    SimulationTime eventTime = event_getTime(event);
+    if (eventTime >= scheduler->endTime) {
+        if (sender == receiver) {
+            warning("Event scheduled for %ld, after end time %ld; freeing", eventTime, scheduler->endTime);
+            event_unref(event);
+        } else {
+            warning("Event scheduled for %ld, after end time %ld; scheduling for dst host to free", eventTime, scheduler->endTime);
```

Comment:
I think this would be pretty common in large tor sims, and doesn't really mean anything went wrong. Should this be just debug level instead?

Reply:
Oops, I thought I was marked as reviewer on this... Nvm :)

Reply:
Yeah I'm still debugging and trying to work this out :). It turns out to be a bit tricky. I was getting detected memory leaks in the CI tor simulations - I think the scheduled "unref" events end up never getting executed sometimes.

I also realized that scheduling them for "now" causes time to potentially move backwards at the destination host. We could schedule for the beginning of the next round, but then again have to worry about weird edge conditions at the end of the simulation.

I'm thinking more about a specialized "drop queue" for each host, or maybe creating an alternate CTask type after all that uses SyncSendPointer instead of HostPointer, which would only be safe to use when the provided pointers are completely "owned" by the task (or otherwise don't need to be guarded by the host lock).